cpp/src/lists/contains.cu

```diff
@@ -44,296 +41,241 @@
-{
-  return &search_key;
-}
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
+template <typename Type>
+auto constexpr is_supported_non_nested_type()
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
-  }
+  return cudf::is_fixed_width<Type>() || std::is_same_v<Type, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
+/**
+ * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
+ *
+ * Depending on the value of `find_first`, a pair of forward or reverse iterators will be returned,
+ * allowing to loop through elements in the list in first-to-last or last-to-first order.
+ *
+ * Note that the element indices always restart to `0` at the first position in each list.
+ *
+ * @tparam find_first A boolean value indicating whether we want to find the first or last
+ *         appearance of a given key in the list.
+ * @param size The number of elements in the list.
+ * @return A pair of {begin, end} iterators to iterate through the range `[0, size)` by forward or
+ *         reverse order.
+ */
+template <bool find_first>
+auto __device__ element_index_pair_iter(size_type const size)
 {
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
+  if constexpr (find_first) {
+    return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
   } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
+    return thrust::pair(thrust::make_reverse_iterator(thrust::make_counting_iterator(size)),
+                        thrust::make_reverse_iterator(thrust::make_counting_iterator(0)));
   }
 }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
-  }
+template <typename Type, typename SearchKeyIter>
+struct search_lists_fn {
+  cudf::detail::lists_column_device_view const lists;
+  SearchKeyIter const keys_iter;
+  duplicate_find_option const find_option;
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+  __device__ thrust::pair<size_type, bool> operator()(size_type const idx) const
```

Comment:
Using a `thrust::optional` would be preferred to a `pair<T, bool>`

Reply:
The returning pair will be written into a zip iterator of `<size_type, bool>`. Can we use `thrust::optional` here for such output?

Reply:
So after looking further at how this is being used, I think this is a rare instance where the use of a `pair<T,bool>` makes sense because of the fact that we're outputting to a zip iterator over two arrays. 

The alternative would be to return an `optional<T>` and add a further transform output iterator to convert an `optional<T>` to a `pair<T,bool>`, but that seems unnecessary.